substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java

```diff
@@ -81,46 +156,132 @@
 
     @Override
     public Stream<DebugDataInfo> dataInfoProvider() {
-        return Stream.empty();
+        return heap.getObjects().stream().filter(this::acceptObjectInfo).map(this::createDebugDataInfo);
     }
 
-    /**
-     * Implementation of the DebugCodeInfo API interface that allows code info to be passed to an
-     * ObjectFile when generation of debug info is enabled.
+    static ObjectLayout OBJECTLAYOUT = ConfigurationValues.getObjectLayout();
+
+    /*
+     * HostedType wraps an AnalysisType and both HostedType and AnalysisType punt calls to
+     * getSourceFilename to the wrapped class so for consistency we need to do type names and path
+     * lookup relative to the doubly unwrapped HostedType.
+     *
+     * However, note that the result of the unwrap on the AnalysisType may be a SubstitutionType
+     * which wraps both an original type and the annotated type that substitutes it. Unwrapping
+     * normally returns the AnnotatedType which we need to use to resolve the file name. However, we
+     * need to use the original to name the owning type to ensure that names found in method param
+     * and return types resolve correctly.
      */
-    private class NativeImageDebugCodeInfo implements DebugCodeInfo {
-        private final HostedMethod method;
-        private final ResolvedJavaType javaType;
-        private final CompilationResult compilation;
+    protected static ResolvedJavaType getJavaType(HostedType hostedType, boolean wantOriginal) {
+        ResolvedJavaType javaType;
+        if (wantOriginal) {
+            // check for wholesale replacement of the original class
+            javaType = hostedType.getWrapped().getWrappedWithoutResolve();
+            if (javaType instanceof SubstitutionType) {
+                return ((SubstitutionType) javaType).getOriginal();
+            } else if (javaType instanceof CustomSubstitutionType<?, ?>) {
+                return ((CustomSubstitutionType<?, ?>) javaType).getOriginal();
+            } else if (javaType instanceof LambdaSubstitutionType) {
+                return ((LambdaSubstitutionType) javaType).getOriginal();
+            } else if (javaType instanceof InjectedFieldsType) {
+                return ((InjectedFieldsType) javaType).getOriginal();
+            } else {
+                return javaType;
+            }
+        }
+        return hostedType.getWrapped().getWrapped();
+    }
+
+    protected static ResolvedJavaType getJavaType(HostedMethod hostedMethod, boolean wantOriginal) {
+        if (wantOriginal) {
+            // check for wholesale replacement of the original class
+            HostedType hostedType = hostedMethod.getDeclaringClass();
+            ResolvedJavaType javaType = hostedType.getWrapped().getWrappedWithoutResolve();
```

Comment:
Consider extracting a new method `ResolvedJavaType getJavaType(ResolvedJavaType, boolean)` to reduce code duplication (i.e. keep all the instanceof checks in one place)

Reply:
Done